libtenzir/src/secret.cpp

```diff
@@ -9,139 +9,371 @@
+      return match(deref(transformed.secret()), self);
+    },
+  };
+  return match(*buffer, f);
 }
 
+namespace {
+
+template <bool prepend>
+auto prepend_append_literal_impl(flatbuffers::FlatBufferBuilder& fbb,
+                                 const fbs::data::Secret& s,
+                                 std::string_view literal) -> secret_offset_t {
+  const auto f = detail::overload{
+    // literal + literal -> literal
+    [&](const fbs::data::SecretLiteral& lit) {
+      if constexpr (prepend) {
+        return make_literal(fbb, literal, deref(lit.value()).string_view());
+      } else {
+        return make_literal(fbb, deref(lit.value()).string_view(), literal);
+      }
+    },
+    // literal + name -> concat
+    // name + literal -> concat
+    [&](const fbs::data::SecretName& name) {
+      auto secrets = concat_offsets_t{};
+      secrets.reserve(2);
+      if constexpr (prepend) {
+        secrets.push_back(make_literal(fbb, literal));
+        secrets.push_back(make_name(fbb, deref(name.value()).string_view()));
+      } else {
+        secrets.push_back(make_name(fbb, deref(name.value()).string_view()));
+        secrets.push_back(make_literal(fbb, literal));
+      }
+      return finish_concatenation(fbb, secrets);
+    },
+    // concat + literal -> concat
+    // literal + concat -> concat
+    // Also performs a check for `concat[0]` or `concat[size-1]` respectively,
+    // to join these.
+    [&](const fbs::data::SecretConcatenation& concat) {
+      concat_offsets_t offsets;
+      if constexpr (prepend) {
+        const auto& secrets = deref(concat.secrets());
+        TENZIR_ASSERT(secrets.size() > 0);
+        const auto& s0 = deref(secrets[0]);
+        if (const auto* lit_0 = try_as<fbs::data::SecretLiteral>(s0)) {
+          offsets.reserve(deref(concat.secrets()).size());
+          const auto lit_0_text = deref(lit_0->value()).string_view();
+          offsets.push_back(make_literal(fbb, literal, lit_0_text));
+          copy_parts(fbb, concat, offsets, 1);
+        } else {
+          offsets.reserve(deref(concat.secrets()).size() + 1);
+          offsets.push_back(make_literal(fbb, literal));
+          copy_parts(fbb, concat, offsets, 0);
+        }
+      } else {
+        const auto& secrets = deref(concat.secrets());
+        TENZIR_ASSERT(secrets.size() > 0);
+        const auto& sn = deref(secrets[secrets.size() - 1]);
+        if (const auto* lit_n = try_as<fbs::data::SecretLiteral>(sn)) {
+          offsets.reserve(deref(concat.secrets()).size());
+          const auto lit_n_text = deref(lit_n->value()).string_view();
+          copy_parts(fbb, concat, offsets, 0, secrets.size() - 1);
+          offsets.push_back(make_literal(fbb, lit_n_text, literal));
+        } else {
+          offsets.reserve(deref(concat.secrets()).size() + 1);
+          copy_parts(fbb, concat, offsets);
+          offsets.push_back(make_literal(fbb, literal));
+        }
```

Comment:
This whole function and the one below is a bit overkill. A simple concat would have sufficed. I'll just trust you that there is no off-by-one or ordering error here.